### PR TITLE
Groundtruth conversion scripts

### PR DESCRIPTION
**Pull request by wmlba**
*Description of changes:*
Adding a notebook that walks through the steps to convert ground truth semantic segmentation manifest from a mask file into a json COCO format

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+   "source": [
+    "import json\n",
+    "import numpy as np\n",
+    "from pycocotools import mask\n",
+    "from skimage import measure\n",
+    "\n",
+    "from skimage import io\n",
+    "\n",
+    "def generate_segmentation(img, category_id, idd, image_id):\n",
+    "    seg_img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)\n",
```
Comment (natsirtguy) [inline ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:183]: You need to import cv2.

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+    "    for idds, seg in enumerate(range(len(segments))):\n",
+    "\n",
+    "        color_hex = matplotlib.colors.to_hex(segments[seg])\n",
+    "        category_id = mapping[color_hex]\n",
+    "        img = io.imread(img_file, plugin='matplotlib')\n",
+    "        msk = segments[seg]  \n",
+    "        masked_img = cv2.inRange(img, msk, msk)\n",
+    "        \n",
+    "        if len(segments) > 1:\n",
+    "            img[masked_img>0]=(255,255,255)\n",
```
Comment (natsirtguy) [inline ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:276]: When I try to run this, this breaks, as img is a 4-channel (RGBA) mask. Maybe you want to do
```suggestion
    "            img[masked_img>0]=(255,255,255,255)\n",
```
instead?

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+    "print('COCO supercategories: \\n{}'.format(' '.join(nms)))\n",
+    "\n",
+    "\n",
+    "catIds = coco.getCatIds(catNms=['bike','motorbike']);\n",
+    "imgIds = coco.getImgIds(catIds=catIds );\n",
+    "imgIds = coco.getImgIds(imgIds = [1]) #<<<<<<<<<<<< Change the number for different test images\n",
+    "img = coco.loadImgs(imgIds[np.random.randint(0,len(imgIds))])[0]\n",
+    "print(img)\n",
+    "\n",
+    "I = io.imread(img['file_name'])\n",
```
Comment (natsirtguy) [inline ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:348]: This doesn't work for me, unless I download the original (unannotated) images into the working directory. Maybe either add the logic to due this by retrieving the input manifest, or at least give a warning to the user?

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+    "from skimage.viewer import ImageViewer\n",
+    "import matplotlib.pyplot as plt\n",
+    "from skimage.color import rgb2gray\n",
+    "\n",
+    "\n",
+    "a = io.imread(base64.decodebytes(img_data), plugin='imageio')\n",
+    "grayscale = rgb2gray(a)\n",
+    "grayscale[grayscale > 0] = 1\n",
+    "plt.imshow(a)\n",
+    "plt.show()"
```
Comment (natsirtguy) [inline ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:418]: Is this cell supposed to be here still? img_data isn't defined, and I'm not sure what's supposed to be going on.

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+    "from pycocotools.coco import COCO\n",
+    "import cv2\n",
+    "\n",
+    "for annotation in annotations:\n",
+    "    with open('COCO_file.json', 'w') as json_file:\n",
+    "        json.dump(annotation, json_file)\n",
+    "    coco=COCO('COCO_file.json')\n",
+    "    I = io.imread('<sample_image>')\n",
+    "    plt.imshow(I)\n",
+    "    coco.showAnns([annotation])"
```
Comment (natsirtguy) [inline ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:399]: Is this supposed to be loading all of the images and plotting the annotations? If so, you might need to make a new figure for each image (or maybe plot out to a pdf), and add logic to get the image ID from the annotation.

**Comment by wmlba** [comment]
Thanks for the review. I will have a look and rectify soon.

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 1 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:183]: Done

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 2 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:276]: @natsirtguy What error do you get? It works fine for me with no issues.

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 1 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:201]: Done

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 3 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb]: Fixed

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 4 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb]: Fixed this one. This is supposed to pick a random image from the dataset (from the output manifest) then overlay the annotation on the image

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 5 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb]: I changed it to just plot one random image but ploting all images in a pdf sounds like a great idea so I can do that as an improvement

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+    "import boto3\n",
+    "\n",
+    "\n",
+    "JOBNAME = '' #Replace it with the labeling job name\n",
+    "REGION = ''\n",
+    "client = boto3.client('sagemaker',region_name=REGION)\n",
+    "\n",
+    "response = client.describe_labeling_job(LabelingJobName= JOBNAME)\n",
+    "file = response['LabelingJobOutput']['OutputDatasetS3Uri']\n",
+    "output_manifest = ntpath.basename(file)\n",
```
[the remark this answers is not in the document]
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb]: all done. I switch to a proper os.path instead

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 2 above)
Comment (natsirtguy) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:285]: Still getting it:
```
<ipython-input-8-ba93f71864dd> in <module>
     24         masked_img = cv2.inRange(img, msk, msk)
     25         if len(segments) > 1:
---> 26             img[masked_img>0]=(255,255,255)
     27         annot = generate_segmentation(img,category_id, idd, image_id)
     28         idd+=1

ValueError: shape mismatch: value array of shape (3,) could not be broadcast to indexing result of shape (1764,4)
```
It's possible that the output format changed from when you ran your labeling job. Did you use one annotator per image for yours?

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+    "        category_id = mapping[color_hex]\n",
+    "        img = io.imread(img_file, plugin='matplotlib')\n",
+    "        msk = segments[seg]  \n",
+    "        masked_img = cv2.inRange(img, msk, msk)\n",
+    "        if len(segments) > 1:\n",
+    "            img[masked_img>0]=(255,255,255)\n",
+    "        annot = generate_segmentation(img,category_id, idd, image_id)\n",
+    "        idd+=1\n",
+    "        annotations.append(annot)\n",
+    "        print(\"Processing image {}:\".format())"
```
Comment (natsirtguy) [inline ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:289]: I think this is missing the image index, and maybe it should go at the top of the loop?

**Review by natsirtguy — CHANGES_REQUESTED** [review]
Almost there.

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 7 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb]: Fixed

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 2 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb]: Fixed

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+    "#Get data objects from the output manifest\n",
+    "with open(file_name) as out_manifest:\n",
+    "    for line in out_manifest:\n",
+    "        data_objs.append(json.loads(line))\n",
+    "\n",
+    "#Get the input images filenames\n",
+    "for line in data_objs:\n",
+    "        input_files.append(os.path.basename(line['source-ref']))\n",
+    "\n",
+    "dir_path = os.path.dirname(data_objs[0]['source-ref'])\n",
```
Comment (natsirtguy) [inline ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:84]: This finds the s3 folder of the first object in the manifest. Later on, you use it to download a test file. However, there is no guarantee that all the objects in the manifest are in the same s3 folder. As a result, the download fails for me. For example, if my first object is at 
`'s3://some-bucket/foo/example1.png'`
and the second is at
`'s3://some-bucket/bar/example2.png'`,
the notebook might try to download `'s3://some-bucket/foo/example2.png'`, which doesn't exist.

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "import matplotlib\n",
+    "from skimage import io\n",
+    "import matplotlib.pyplot as plt\n",
+    "from skimage.color import rgb2gray\n",
+    "\n",
+    "idd = 0\n",
+    "all_annotations = []\n",
```
Comment (natsirtguy) [inline ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:267]: `all_annotations` is unused.

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+    "width = 5280\n",
+    "\n",
+    "#Get data objects from the output manifest\n",
+    "with open(file_name) as out_manifest:\n",
+    "    for line in out_manifest:\n",
+    "        data_objs.append(json.loads(line))\n",
+    "\n",
+    "#Get the input images filenames\n",
+    "for line in data_objs:\n",
+    "        input_files.append(os.path.basename(line['source-ref']))\n",
```
Comment (natsirtguy) [inline ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:82]: Similar to the issue below: You're assuming that all the annotated images have distinct file basenames. If you have something like `'s3://some-bucket/foo/0.png'` and `'s3://some-bucket/bar/0.png'` in the same output manifest, both are going to be put into COCO_file.json with the same `'file_name'` attribute, `'0.png'`, and you won't be able to tell which annotation goes with which image.

**Review by natsirtguy — CHANGES_REQUESTED** [review]
I realized there are some issues with just using the file basenames. I think everything should work if you just use the full `source-ref` without the bucket name instead.

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 10 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb]: resolved

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 9 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb]: removed it

File: ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb (hunk 8 above)
Comment (wmlba) [reply ground_truth_labeling_jobs/ground_truth_conversion_scripts/GT_semantic_segmentation_to_COCO.ipynb:85]: This is resolved by adding the path to the json file and preserving the path in the local dir as well

**Review by natsirtguy — DISMISSED** [review]
LGTM, thanks for the hard work!

**Review by TEChopra1000 — APPROVED** [review]

**Pull request merged**